### ISSUE TITLE
docker cp --follow-link does not work as expected

### ISSUE THREAD

**Issue by josh-tepper** [docker/docker#30213]
**Description**

There's one other ambiguity in the documentation/implementation that I've come across. When I run:

`docker cp --follow-link container_id:source/path dest/path`

I would expect that when `source/path/` contains files that are symlinks, e.g. `source/path/file.symlink`, and subdirectories that are symlinks, e.g., `souce/path/dir.symlink/more_files`, that all of those would be resolved, and that I would get all the files at dest/path. This is not the case. Instead I end up with broken symlinks.

My best guess is that --follow-link refers to links that are in source path itself, e.g., `souce.symlink/path/`, but I haven't tested that. Even so, that doesn't really make so much sense to me as a feature. This is because as @jlhawn pointed out in issue https://github.com/docker/docker/issues/30082, there is already a *nix cp notion as to whether to resolve symlinks at the end of path (based on the trailing /), and it seems that symlinks in the middle of a path should always be resolved.

**Steps to reproduce the issue:**
1. run `docker --follow-link container_id:source/path dest/path` where source path contains symlinked files/subdir
2. The files at `dest/path` will contain broken symlinks

**Output of `docker version`:**

```
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Wed Jan 11 00:23:16 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Wed Jan 11 00:23:16 2017
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 11
 Running: 0
 Paused: 0
 Stopped: 11
Images: 52
Server Version: 1.12.6
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 98
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.41-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.951 GiB
Name: moby
ID: XEI4:KV5E:QUMI:STWS:2Y2X:MXS2:X5QC:FUHS:Z27W:ZW3I:4ZKO:O4KR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 17
 Goroutines: 29
 System Time: 2017-01-12T00:03:34.814396512Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running on Mac OS 10.11.6

**Comment by jlhawn** [comment]
> My best guess is that --follow-link refers to links that are in source path itself

Yes, this is what the flag is meant to do. Unfortunately there is currently no option to follow symlink targets which exist in a source directory or its subdirectories when copying them. Doing so would probably require changes to the underlying archive utility which `docker cp` uses under the hood as well as special logic to detect and avoid symlinked directory cycles which the user might have created.

**Comment by trevor-vaughan** [comment]
Just ran into this and the current behavior essentially renders `docker cp` relatively useless a LOT of cases. I think I can fall back to `tar` pipes but that's not awesome.

**Comment by ptr727** [comment]
I just encountered the same problem, I wanted to get copies of fresh nginx config files from Ubuntu and docker, and -L does not follow symlinks as expected.

docker cp -L nginxtmp:/etc/nginx/. ./etcnginx/
docker cp -L nginxtmp:/usr/share/nginx/. ./usrsharenginx/
docker cp -L nginxtmp:/usr/lib/nginx/. ./usrlibnginx/

cp -LR /etc/nginx/. ./etcnginx/
cp -LR /usr/share/nginx/. ./usrsharenginx/
cp -LR /usr/lib/nginx/. ./usrlibnginx/

cp -LR correctly resolves the modules link in etc/nginx to usr/lib/nginx/modules, while cp -L does not.
